packages/eas-cli/src/credentials/ios/IosCredentialsProvider.ts

```diff
@@ -143,17 +143,32 @@ export default class IosCredentialsProvider {
 
   private assertProvisioningProfileType(provisioningProfile: string, targetName?: string): void {
     const isAdHoc = isAdHocProfile(provisioningProfile);
-    if (this.options.distribution === 'internal' && !isAdHoc) {
-      throw new Error(
-        `You must use an adhoc provisioning profile${
-          targetName ? ` (target '${targetName})'` : ''
-        } for internal distribution`
-      );
-    } else if (this.options.distribution !== 'internal' && isAdHoc) {
+    const isEnterprise = isEnterpriseUniversalProfile(provisioningProfile);
+    if (this.options.distribution === 'internal') {
+      if (this.options.enterpriseProvisioning === 'universal' && !isEnterprise) {
+        throw new Error(
+          `You must use an universal provisioning profile${
+            targetName ? ` (target '${targetName})'` : ''
+          } for internal distribution if you specified "enterpriseProvisioning": "universal" in eas.json`
+        );
+      } else if (this.options.enterpriseProvisioning === 'adhoc' && !isAdHoc) {
+        throw new Error(
+          `You must use an adhoc provisioning profile${
+            targetName ? ` (target '${targetName})'` : ''
+          } for internal distribution if you specified "enterpriseProvisioning": "adhoc" in eas.json`
+        );
+      } else if (!this.options.enterpriseProvisioning && !isEnterprise && !isAdHoc) {
+        throw new Error(
+          `You must use an adhoc provisioning profile${
```

Comment:
Most user is not aware what enterprise accounts, so decided to mention only ad hoc here, but I'm open to other suggestions.